File: workflow-service/src/execution/instance/instance.service.ts
import { Injectable } from '@nestjs/common';
import { InstanceId, Instance } from './instance';
import { v4 as uuid } from 'uuid';

@Injectable()
export class InstanceService {
  instances: Map<InstanceId, Instance> = new Map();

  findInstance(instanceId: InstanceId): Instance {
    const instance = this.instances.get(instanceId);
    if (!instance) {
      throw Error(`Instance ${instanceId} not found`);
    }
    return instance;
  }

  saveInstance(instance: Instance) {
    if (!instance.id) {
      instance.id = this.createInstanceId();
    }
    this.instances.set(instance.id, instance);
  }

  private createInstanceId(): InstanceId {
    return uuid();
  }
}

File: execution-service/domain/instance_service.go
package domain

import (
	"fmt"
)

type InstanceService struct {
	isLoaded  bool
	instances map[string]Instance
}

var instanceService InstanceService

func NewInstanceService() InstanceService {
	if !instanceService.isLoaded {
		instanceService = InstanceService{
			isLoaded:  true,
			instances: make(map[string]Instance),
		}
	}
	return instanceService
}

func (service *InstanceService) SaveInstance(instance Instance) {
	service.instances[instance.Id] = instance
}

func (service *InstanceService) FindInstanceById(id InstanceId) (Instance, error) {
	instance, exists := service.instances[id]
	if !exists {
		return Instance{}, fmt.Errorf("instance %s not found", id)
	}
	return instance, nil
}

func (service *InstanceService) FindInstancesByModel(model ModelId) []Instance {
	instances := make([]Instance, 0, len(service.instances))
	for _, instance := range service.instances {
		if instance.Model == model {
			instances = append(instances, instance)
		}
	}
	return instances
}
